cumulus/client/consensus/aura/src/collator.rs

```diff
@@ -433,3 +441,130 @@ where
+	spawn_handle: Spawner,
+) where
+	Block: sp_runtime::traits::Block,
+	Client: sc_client_api::HeaderBackend<Block>
+		+ Send
+		+ Sync
+		+ ProvideRuntimeApi<Block>
+		+ sc_client_api::BlockchainEvents<Block>
+		+ 'static,
+	Client::Api: AuraApi<Block, P::Public>,
+	P: sp_core::Pair + Send + Sync,
+	P::Public: Codec,
+	Spawner: sp_core::traits::SpawnNamed,
+{
+	let mut import_notifications = client.import_notification_stream().fuse();
+
+	tracing::debug!(target: crate::LOG_TARGET, "Started collator protocol helper");
+
+	// Our own slot number, known in advance.
+	let mut our_slot = None;
+
+	while let Some(notification) = import_notifications.next().await {
+		// Determine if this node is the next author
+		let digest = notification.header.digest();
+		let slot = digest
+			.logs()
+			.iter()
+			.find_map(|log| <DigestItem as CompatibleDigestItem<Slot>>::as_aura_pre_digest(log));
+
+		tracing::debug!(target: crate::LOG_TARGET, "Imported block with slot: {:?}", slot);
+
+		let Some(slot) = slot else {
+			continue;
+		};
+
+		// Determine current slot duration.
+		let Ok(slot_duration) = client.runtime_api().slot_duration(notification.header.hash())
+		else {
+			continue;
+		};
+
+		let authorities =
+			client.runtime_api().authorities(notification.header.hash()).unwrap_or_default();
+
+		// Check if our slot has passed and we are not expected to author again in next slot.
+		match (
+			our_slot,
+			aura_internal::claim_slot::<P>(slot + 1, &authorities, &keystore)
+				.await
+				.is_none(),
+		) {
+			(Some(last_slot), true) if slot > last_slot => {
+				tracing::debug!(target: crate::LOG_TARGET, "Our slot {} has passed, current slot is {}, sending disconnect message", last_slot, slot);
+
+				// Send a message to the collator protocol to stop pre-connecting to backing
+				// groups
+				overseer_handle
+					.send_msg(
+						CollatorProtocolMessage::DisconnectFromBackingGroups,
+						"CollatorProtocolHelper",
+					)
+					.await;
+
+				our_slot = None;
+			},
+			_ => {},
+		}
+
+		// Check if our slot is coming up next. This means that there is still another slot
+		// before our turn.
+		let target_slot = slot + 2;
+		if aura_internal::claim_slot::<P>(target_slot, &authorities, &keystore)
+			.await
+			.is_none()
+		{
+			continue;
+		}
+
+		tracing::debug!(target: crate::LOG_TARGET, "Our slot {} is due soon", target_slot );
+
+		// Determine our own slot timestamp.
+		let Some(own_slot_ts) = target_slot.timestamp(slot_duration) else {
+			tracing::warn!(target: crate::LOG_TARGET, "Failed to get own slot duration");
+
+			continue;
+		};
+
+		let pre_connect_delay = own_slot_ts
+			.saturating_sub(*Timestamp::current())
+			.saturating_sub(PRE_CONNECT_SLOT_OFFSET.as_millis() as u64);
+
+		tracing::debug!(target: crate::LOG_TARGET, "Pre-connecting to backing groups in {}ms", pre_connect_delay);
+
+		let mut overseer_handle_clone = overseer_handle.clone();
+		spawn_handle.spawn(
```

Comment:
If we have already sent a connection request for a target slot to the backing group, lets check that and not send again. When there is many small blocks in the same slot, we would currently send the message multiple times.